making sidecar scc compliant with restricted PSS

JaydipGabani opened the pull request:
# Description
Right now there is no way to inject appropriate `securityContext` in a sidecar container that can run in a namespace with restricted PSS labels. For example, dropping all capabilities is a must for a container to run in a namespace with restricted PSS labels, but there are no settings for injecting the same through dapr-sidecar-injector. This PR introduces two configurable options to set appropriate `securityContext`.

<!--
Please explain the changes you've made.
-->

## Issue reference

<!--
We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.
-->

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

yaron2 commented on pkg/injector/sidecar/container.go:
```diff
@@ -194,6 +194,19 @@ func GetSidecarContainer(cfg ContainerConfig) (*corev1.Container, error) {
 		cfg.DaprSidecarImage = image
 	}
 
+	// securityContext := &corev1.SecurityContext{
```
Please remove commented code

ItalyPaleAle commented on pkg/injector/annotations/annotations.go:
```diff
@@ -47,6 +47,9 @@ const (
 	KeyReadinessProbeThreshold          = "dapr.io/sidecar-readiness-probe-threshold"
 	KeySidecarImage                     = "dapr.io/sidecar-image"
+	KeySidecarProfile                   = "dapr.io/sidecar-profile"
+	KeySidecarDropCapability            = "dapr.io/sidecar-drop-capability"
```
should this be `sidecar-drop-capabilities`?

JaydipGabani replied:
Base on the comment [here](https://github.com/dapr/dapr/issues/5916#issuecomment-1433879612), I removed both the annotations and implemented a way to drop or preserve capabilities through helm chart

ItalyPaleAle commented on pkg/injector/annotations/annotations.go:
```diff
@@ -47,6 +47,9 @@ const (
 	KeyReadinessProbePeriodSeconds      = "dapr.io/sidecar-readiness-probe-period-seconds"
 	KeyReadinessProbeThreshold          = "dapr.io/sidecar-readiness-probe-threshold"
 	KeySidecarImage                     = "dapr.io/sidecar-image"
+	KeySidecarProfile                   = "dapr.io/sidecar-profile"
```
Can you be more specific on what kind of profile this is? "sidecar-profile" is not immediately clear without context

JaydipGabani replied:
Yeah, removed this annotation as it might be confusing.

ItalyPaleAle commented on pkg/injector/sidecar/container.go:
```diff
@@ -202,6 +215,8 @@ func GetSidecarContainer(cfg ContainerConfig) (*corev1.Container, error) {
 			AllowPrivilegeEscalation: ptr.Of(false),
 			RunAsNonRoot:             ptr.Of(cfg.RunAsNonRoot),
 			ReadOnlyRootFilesystem:   ptr.Of(cfg.ReadOnlyRootFilesystem),
+			Capabilities:             &corev1.Capabilities{Drop: []corev1.Capability{"ALL"}},
```
This should be set only if the annotation wants this, right? Same for the one below?

mukundansundar commented:
@JaydipGabani Could you add more details on what this PR tries to resolve? 
Also is there an issue linked with this change?

JaydipGabani commented:
> @JaydipGabani Could you add more details on what this PR tries to resolve? Also is there an issue linked to this change?

Sorry for the delay in updating the description and the need for this PR, but now it's there. 

There is no issue, but we deduced the requirement from #5916

JaydipGabani commented:
@ItalyPaleAle @yaron2 sorry for the delay. But I have addressed your feedback and improved the implementation as well. Please take a look again.

ItalyPaleAle commented on charts/dapr/charts/dapr_sidecar_injector/templates/dapr_sidecar_injector_deployment.yaml:
```diff
@@ -147,6 +147,8 @@ spec:
 {{- end }}
         - name: SIDECAR_RUN_AS_NON_ROOT
           value: "{{ .Values.sidecarRunAsNonRoot }}"
+        - name: DROP_CAPABILITIES
```
Can you change this to `SIDECAR_DROP_CAPABILITIES` for consistency?

ItalyPaleAle replied:
Or maybe even `SIDECAR_DROP_ALL_CAPABILITIES`

JaydipGabani replied:
Good point! I updated the name for consistency

ItalyPaleAle reviewed (changes requested):
Thanks, It looks good to me just changing the env var name please. I'd also suggest changing the names everywhere to indicate this is for dropping "all" capapbilities

yaron2 commented on charts/dapr/charts/dapr_sidecar_injector/values.yaml:
```diff
@@ -22,6 +22,7 @@ sidecarImagePullPolicy: IfNotPresent
 runAsNonRoot: true
 sidecarRunAsNonRoot: true
 sidecarReadOnlyRootFilesystem: true
+sidecarDropALLCapabilities: true
```
This should default to false.

ItalyPaleAle commented:
@JaydipGabani LGTM but please don't rebase your commits. That breaks things like the ability to view changes since the last review, so it forces reviewers to have to review from scratch again.

Most of our PRs get squashed when we merged them in master anyways :)

JaydipGabani commented:
> @JaydipGabani LGTM but please don't rebase your commits. That dismisses approvals and breaks things like the ability to view changes since the last review, so it forces reviewers to have to review from scratch again.
> 
> Most of our PRs get squashed when we merged them in master anyways :)

Ohh My bad! I will be mindful about this

yaron2 commented on pkg/injector/config.go:
```diff
@@ -105,6 +106,14 @@ func (c *Config) GetReadOnlyRootFilesystem() bool {
+func (c *Config) GetDropCapabilities() bool {
+	// Default is true if empty
+	if c.SidecarDropALLCapabilities == "" {
+		return true
```
Should be `false`. This entire feature should be opt-in, otherwise we run the risk of breaking many systems in many places.

JaydipGabani commented:
@yaron2 @ItalyPaleAle is there an issue with CI? I see same check is failing in other PRs as well

ItalyPaleAle commented:
@JaydipGabani yeah, some of our tests are flaky :( I re-triggered it

yaron2 commented:
@JaydipGabani to complete this feature please open a docs PR to add the required information.
